bazel/PPROF.md

```diff
@@ -18,80 +16,75 @@ Build the static binary using bazel:
-binary must be run with this environment variable; a dynamically-linked binary
-will populate the working directory by default. (See *Methodology*.)
+    $ CPUPROFILE=/tmp/mybin.cpuprof bazel-bin/source/exe/envoy-static <args>
+    $ HEAPPROFILE=/tmp/mybin.heapprof bazel-bin/source/exe/envoy-static <args>
 
-## Dynamic Linking
+`CPUPROFILE` or `HEAPPROFILE` sets a location for the profiler output. (See *Methodology*.)
 
-### Adding `tcmalloc_dep` to Envoy
+There are several other environment variables that can be set to tweak the behavior of gperftools. See https://gperftools.github.io/gperftools/ for more details.
 
-A statically-linked Envoy will profile everything. In a dynamically-linked
-Envoy, you must add the HeapProfiler instructions yourself.
-`HeapProfilerStart()` will start recording allocations, `HeapProfilerStop()`
-will stop recording, and `HeapProfilerDump()` will dump an output to the
-specified directory. (See [Gperftools Heap
-Profiler](https://gperftools.github.io/gperftools/heapprofile.html).)
+### Analyzing the profile
 
-To add a `HeapProfiler` breakpoint yourself, add `tcmalloc` as a
-dependency under the `envoy_cc_library` rule:
+[pprof](https://github.com/google/pprof) can be used to symbolize CPU and heap profiles. For example:
 
-`source/exe/BUILD`
+    $ pprof -text bazel-bin/source/exe/envoy-static /tmp/mybin.cpuprof
 
-```c++
-    envoy_cc_library(
-       name = "envoy_common_lib",
-+      tcmalloc_dep = 1,
-       deps = [
-       ...
-    )
-```
+## Collecting CPU or heap profile for the full execution of a test target
 
-It is then necessary to add `HeapProfilerStart()` and `HeapProfilerDump()`
-breakpoints somewhere in Envoy. One place to start profiling is at the
-instantiation of `MainCommonBase::MainCommonBase`:
+The profiler library is automatically linked into envoy_cc_test targets.
 
-`source/exe/main_common.cc`
+Run a test with CPU profiling enabled, like so:
 
-```c++
-    // includes
-    #include "gperftools/heap-profiler.h"
-    ...
-    MainCommonBase::MainCommonBase(...) : ... {
-+       HeapProfilerStart("main_common_base"); // first line
-        ...
-    }
-```
+    $ bazel test --test_env=CPUPROFILE=/tmp/cpuprof <test target>
+
+Run a test with heap profiling enabled, like so:
+
+    $ bazel test --test_env=HEAPPROFILE=/tmp/heapprof <test target>
+
+## Starting and stopping profile programmatically
 
-`source/server/server.cc`
+### Add `tcmalloc_dep` dependency to envoy_cc_library rules
+
+It is possible to start/stop the CPU or heap profiler programmatically.
+The [Gperftools CPU Profiler](https://gperftools.github.io/gperftools/cpuprofile.html)
+is controlled by `ProfilerStart()`/`ProfilerStop()`, and the
+[Gperftools Heap Profiler](https://gperftools.github.io/gperftools/heapprofile.html)
+is controlled by `HeapProfilerStart()`, `HeapProfilerStop()` and `HeapProfilerDump()`.
+
+These functions are wrapped by Envoy objects defined in source/common/profiler/profiler.h
```

Comment:
Do you want to link to this file in the GH repository?

Reply:
Done.